### PR TITLE
refactor using go client

### PR DESCRIPTION
**Pull request by rawdaGastan**
## Issues

- https://github.com/threefoldtech/terraform-provider-grid/issues/554

tests to check:
- qsfs
- wg
- k8s with modules

### CODE REVIEW

File: internal/provider/deployment.go [hunk 1]
```diff
@@ -2,442 +2,145 @@
...
 	vms := make([]workloads.VM, 0)
 	for _, vm := range d.Get("vms").([]interface{}) {
-		v := workloads.NewVMFromSchema(vm.(map[string]interface{})).WithNetworkName(networkName)
+		vmMap := vm.(map[string]interface{})
+		vmMap["network_name"] = networkName
+		v := workloads.NewVMFromMap(vmMap)
 		vms = append(vms, *v)
 	}
 
+	// TODO: ip_range
```
Comment (ashraffouda) [inline internal/provider/deployment.go:39]: remove unneeded code plz

File: internal/provider/gateway_name.go [hunk 2]
```diff
@@ -2,114 +2,85 @@
...
-	deploymentData := DeploymentData{
-		Name:        d.Get("name").(string),
-		Type:        "gateway",
-		ProjectName: d.Get("solution_type").(string),
-	}
-	deploymentDataStr, err := json.Marshal(deploymentData)
-	if err != nil {
-		log.Printf("error parsing deploymentdata: %s", err.Error())
+
+	var contractID uint64
```
Comment (ashraffouda) [inline internal/provider/gateway_name.go:34]: same as above let's remove unnecessary if statements

File: internal/provider/gateway_name_test.go [hunk 3]
```diff
@@ -1,643 +0,0 @@
-// Package provider is the terraform provider
```
Comment (ashraffouda) [inline internal/provider/gateway_name_test.go:1]: why did we remove tests?

File: internal/provider/k8s.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+	// remove old ips
+	network.DeleteDeploymentHostIDs(k8s.Master.Node, k8s.NodeDeploymentID[k8s.Master.Node])
+	for _, worker := range k8s.Workers {
+		network.DeleteDeploymentHostIDs(worker.Node, (k8s.NodeDeploymentID[worker.Node]))
+	}
+
+	// append new ips
+	masterNodeDeploymentHostIDs := network.GetDeploymentHostIDs(k8s.Master.Node, k8s.NodeDeploymentID[k8s.Master.Node])
+	masterIP := net.ParseIP(k8s.Master.IP)
+	if masterIP == nil {
```
Comment (ashraffouda) [inline internal/provider/k8s.go:145]: is master doesn't have valid ip this is an issue, no?

File: internal/provider/provider.go [hunk 5]
```diff
@@ -142,141 +137,31 @@ func New(version string, st state.Getter) (func() *schema.Provider, subi.Substra
...
-			if !ok {
-				return nil, diag.Errorf("error getting relay url for network %s", network)
-			}
+		timeout := d.Get("rmb_timeout").(int)
+		verifyReply := d.Get("verify_reply").(bool)
+		debug := true
 
-			cl, err = direct.NewClient(context.Background(), keyType, threefoldPluginClient.mnemonics, relayURL, sessionID, sub, true)
-		}
+		// TODO:
```
Comment (ashraffouda) [inline internal/provider/provider.go:153]: let's remove those comments

**Review by ashraffouda — CHANGES_REQUESTED** [review]

**Review by MarioBassem — DISMISSED** [review]
access nodes are not deleted while destroying networks.
i took a look at the code and found that only the contracts on the user-specified nodes are deleted when deleting a network resource, which i think is not wanted behavior???

**Review by AbdelrahmanElawady — COMMENTED** [review]
looks good, some minor comments

**Review by xmonader — COMMENTED** [review]
I'm very happy with the PR, It's huge step forward!

File: internal/provider/k8s.go [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+	}
+
+	nodeDeploymentIDIf := d.Get("node_deployment_id").(map[string]interface{})
+	nodeDeploymentID := make(map[uint32]uint64)
+	for node, id := range nodeDeploymentIDIf {
+		nodeInt, err := strconv.ParseUint(node, 10, 32)
+		if err != nil {
+			return nil, errors.Wrap(err, "couldn't parse node id")
+		}
+		deploymentID := uint64(id.(int))
```
[the remark this answers is not in the document]
Comment (rawdaGastan) [reply internal/provider/k8s.go:37]: terraform doesn't accept uint64 interface

File: internal/provider/k8s.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+	masterNodeDeploymentHostIDs := network.GetDeploymentHostIDs(k8s.Master.Node, k8s.NodeDeploymentID[k8s.Master.Node])
+	masterIP := net.ParseIP(k8s.Master.IP)
+	if masterIP == nil {
+		log.Printf("couldn't parse master ip")
+	} else {
+		masterNodeDeploymentHostIDs = append(masterNodeDeploymentHostIDs, masterIP.To4()[3])
+	}
+	network.SetDeploymentHostIDs(k8s.Master.Node, k8s.NodeDeploymentID[k8s.Master.Node], masterNodeDeploymentHostIDs)
+	for _, worker := range k8s.Workers {
+		workerNodeDeploymentHostIDs := network.GetDeploymentHostIDs(worker.Node, k8s.NodeDeploymentID[worker.Node])
```
[the remark this answers is not in the document]
Comment (rawdaGastan) [reply internal/provider/k8s.go:162]: no as we set them per node

File: internal/provider/k8s.go [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+	k8s := workloads.K8sCluster{
+		Master:           &master,
+		Workers:          workers,
+		Token:            d.Get("token").(string),
+		SSHKey:           d.Get("ssh_key").(string),
+		NetworkName:      d.Get("network_name").(string),
+		SolutionType:     d.Get("solution_type").(string),
+		NodeDeploymentID: nodeDeploymentID,
+		NodesIPRange:     nodesIPRange,
```
[the remark this answers is not in the document]
Comment (rawdaGastan) [reply internal/provider/k8s.go:49]: its calculated through deployment

**Review by xmonader — APPROVED** [review]

**Pull request merged**